engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,184 @@
+
+### EthereumTraceCache
+
+The implementation introduces the following trait, which is implemented primarily by `Store`.
+
+```rust
+use std::ops::Range;
+struct TracesInRange {
+    range: Range<u64>,
+    traces: Vec<Trace>,
+}
+
+pub trait EthereumTraceCache: Send + Sync + 'static {
+    /// Attempts to retrieve traces from the cache. Returns ranges which were retrieved.
+    /// The results may not cover the entire range of blocks. It is up to the caller to decide
+    /// what to do with ranges of blocks that are not cached.
+    fn get_cache(contract_address: Option<H160>, blocks: Range<u64>
+        ) -> Box<dyn Future<Output=Result<Vec<TracesInRange>, Error>>>;
+    fn append_cache(contract_address: Option<H160> traces: Vec<TracesInRange>);
+}
+```
+
+#### Block schema
+
+Each cached block will exist as its own row in the database. Here is the schema for that cache:
+
+```rust
+  id -> (Nullable<Bytea>, Integer), // (contract_address, block_number)
+  trace -> Jsonb // Or Bytea
+```
+
+Traces will be stored as Jsonb or Bytea using MessagePack, whichever is easier.
+
+#### Valid Cache Range
+Because of the absence of trace data for a block is a valid cache result, the database must maintain a data structure indicating which ranges of the cache are valid.
+
+This is the schema for that structure:
+```rust
+start_block -> Integer
+end_block -> Integer
+contract_address -> Nullable<Bytea>
+```
+
+When inserting data into the cache, removing data from the cache, or reading the cache, a serialized transaction must be used to preserve atomicity between the valid cache range structure and the cached blocks. Care must be taken to not rely on any data read outside of the serialized transaction, and for the extent of the serialized transaction to not span any async contexts that rely on any `Future` outside of the database itself. The definition of the `EthereumTraceCache` trait is designed to uphold these guarantees.
+
+In order to preserve space in the database, whenever the valid cache range is added it will be added such that adjacent and overlapping ranges are merged into it.
+
+### Cache usage
+The primary user of the cache is `EtheriumAdapter<T>` in the `traces` function. 
+
+The correct algorithm for retrieving traces from the cache is surprisingly nuanced. The complication arises from the interaction between multiple subgraphs which may require a subset of overlapping contract addresses. The rate at which indexing proceeds of these subgraphs can cause different ranges of the cache to be valid for a contract address in a single query.
+
+We want to minimize the cost of external requests for trace data. It is likely that it is better too...
+ * Make fewer requests
+ * Not ask for trace data that is already cached
+ * Ask for trace data for multiple contract addresses within the same block when possible.
```

Comment:
My understanding is that our `fn trace` has a parameter `addresses: Vec<H160>` which sometimes maps to `toAddresses` in this API, which takes an array.

About "sometimes maps to" - I'd like to change this to an `Option<Vec<_>>` to make it more clear, rather than having an empty `Vec` have a magic meaning.

Reply:
Yeah, that makes sense (regarding the `Option<Vec<_>>`). It's always good to be explicit.